Add blog post about Micronaut Test Resources (#12)
melix/blog: src/jbake/content/2022/07/micronaut-test-resources.adoc

:jbake-type: post
:jbake-tags: micronaut, testcontainers, docker, test, testing
:jbake-status: published
:source-highlighter: pygments
:id: micronaut-test-resources-draft
:linkattrs:
:mtr: https://micronaut-projects.github.io/micronaut-test-resources/latest/guide/

The new https://micronaut.io/2022/05/26/micronaut-framework-3-5-0/[release of Micronaut 3.6] introduces a new feature which I worked on for the past couple of months, called {mtr}[Micronaut Test Resources].
This feature, which is inspired from https://quarkus.io/guides/dev-services[Quarkus' Dev Services], will greatly simplify testing of Micronaut applications, both on the JVM and using GraalVM native images.
Let's see how.

== Test resources in a nutshell

{mtr}[Micronaut Test Resources] simplifies testing of applications which depend on external resources, by handling the provisioning and lifecycle of such resources automatically.
For example, if your application requires a MySQL server, in order to test the application, you need a MySQL database to be installed and configured, which includes a database name, a username and a password.
In general, those are only relevant for production, where those are fixed.

> During development, all you care about is to have _one_ database available.

Reviewer: ```suggestion
During development, all you care about is having _one_ database available.
```

[source]
----
INFO  io.micronaut.runtime.Micronaut - Startup completed in 1086ms. Server Running: http://localhost:8080
----

This time the application started in just a second! Let's add a book:

[source,bash]
----
$ http :8080/books/add/Micronaut%20in%20action
HTTP/1.1 200 OK
Content-Type: application/json
connection: keep-alive
content-length: 38
date: Tue, 26 Jul 2022 17:03:57 GMT

{
    "id": 1,
    "title": "Micronaut in action"
}

----

However, if we stop the application (by hitting CTRL+C) and start again, you will see that the database will be destroyed when the application shuts down.
Let's see how we can "survive" different build invocations.

=== Keeping the service alive

By default, the test resources service is _short lived_: it's going to be started at the beginning of a build, and shutdown at the end of a build.
This means, that it will live as long as you have tests running, or, if running in development mode, as long as the application is alive.
However, you can make it survive the build, and reuse the containers in several, independent build invocations.

To do this, you need to _explicitly start the test resources service_:

> `./gradlew startTestResourcesService`

Reviewer: Is there a maven analog of this?

Author: Yes, I will add it :+1:

[source]
----
INFO  io.micronaut.runtime.Micronaut - Startup completed in 9211ms. Server Running: http://localhost:8080
----

We can add our book:

[source,bash]
----
$ http :8080/books/add/Micronaut%20in%20action
HTTP/1.1 200 OK
Content-Type: application/json
connection: keep-alive
content-length: 38
date: Tue, 26 Jul 2022 17:03:57 GMT

{
    "id": 1,
    "title": "Micronaut in action"
}

----

The difference is now that if we stop the application (e.g hit CTRL+C) and start it again, it will _reuse the container_:

[source]
----
INFO  io.micronaut.runtime.Micronaut - Startup completed in 895ms. Server Running: http://localhost:8080
----

If we list our books, the database wasn't cleaned, so we'll get the book we created from the previous time we started the app:

[source,bash]
----
$ http :8080/books
HTTP/1.1 200 OK
Content-Type: application/json
connection: keep-alive
content-length: 40
date: Tue, 26 Jul 2022 17:14:40 GMT

[
    {
        "id": 1,
        "title": "Micronaut in action"
    }
]
----

Nice, right? 
However there's a gotcha if you do this: what happens if we run tests?

[source,bash]
----

> $ gw test

Reviewer: ```suggestion
$ ./gradlew test
```
